src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -6080,6 +6080,25 @@ public void Invoke(Task completingTask)
         /// </exception>
         public static Task<Task> WhenAny(IEnumerable<Task> tasks)
         {
+            // Take a more efficient path if tasks is actually an array
+            if (tasks is Task[] taskArray)
+            {
+                return WhenAny(taskArray);
+            }
+
+            // Skip a List allocation/copy if tasks is a collection
+            if (tasks is ICollection<Task> taskCollection)
+            {
+                int index = 0;
+                taskArray = new Task[taskCollection.Count];
+                foreach (Task task in tasks)
+                {
+                    if (task == null) ThrowHelper.ThrowArgumentException(ExceptionResource.Task_MultiTaskContinuation_NullTask, ExceptionArgument.tasks);
+                    taskArray[index++] = task;
+                }
+                return TaskFactory.CommonCWAnyLogic(taskArray);
+            }
```

Comment:
This seems reasonable.

Technically this is introducing failure conditions that didn't previously exist, e.g. if the supplied enumerable is some kind of concurrent collection such that its Count could change after accessing it, this could throw an exception where it didn't previously.  But that's dubious to begin with, and the fact that the corresponding WhenAll implementation already does this exact set of checks suggests it's fine to bring them in line with each other.

There's also the two additional cast operations that'll be incurred on the slow path, but the slow path already has non-trivial overheads, and it's unlikely those two cast checks will matter in practice. That said, while it's very likely an `ICollection<Task>` is passed in here, it seems a lot less likely that a `Task[]` is passed in, as if the argument were actually typed as `Task[]`, it would bind to a different overload.  So it would need to be coming from some helper method that returned a `Task[]` typed as `IEnumerable<Task>`, and that's not too common in my experience with `Task.WhenAny`.  I'd suggest moving the first if block to be inside of the second if block:
```C#
if (tasks is ICollection<Task> taskCollection)
{
    if (tasks is Task[] taskArray)
    {
        ...
    }

    ...
}
```

Reply:
Sounds good, should I change `WhenAll` too? So the behavior is the same

Reply:
Yeah, I think that's reasonable.